### PR TITLE
Metro volume expand support

### PR DESCRIPTION
**Pull request by aqu-dell**
# Description

- When we expand one of the metro volumes, we want the other to expand as well in response. In these changes we, pause the replication session before we expand the local volume, and then resume after we expand the local volume. The remote volume should expand as a result.
- Note: We observed that _NodeExpandVolume_ will continue to fail till the metro session is resumed and remote volume gets updated size.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/1443 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Tested expanding both regular and metro volume, and they succeeded.
Controller and node logs are attached to ticket 26867.

Node log for metro volume:
```
"/csi.v1.Node/NodeExpandVolume"
{"level":"info","msg":"Executing rescan command on device (/sys/block/sdm)","time":"2024-10-04T19:11:20.957901125Z"}
{"level":"debug","msg":"Rescan output","output":"","time":"2024-10-04T19:11:21.3519596Z"}
{"level":"info","msg":"Successful rescan on device (/sys/block/sdm)","time":"2024-10-04T19:11:21.351998885Z"}
{"level":"info","msg":"Executing rescan command on device (/sys/block/sdn)","time":"2024-10-04T19:11:21.352014214Z"}
{"level":"debug","msg":"Rescan output","output":"","time":"2024-10-04T19:11:21.876234352Z"}
{"level":"info","msg":"Successful rescan on device (/sys/block/sdn)","time":"2024-10-04T19:11:21.876306324Z"}
{"level":"info","msg":"Executing rescan command on device (/sys/block/sdl)","time":"2024-10-04T19:11:21.876323373Z"}
{"level":"debug","msg":"Rescan output","output":"","time":"2024-10-04T19:11:22.255125618Z"}
{"level":"info","msg":"Successful rescan on device (/sys/block/sdl)","time":"2024-10-04T19:11:22.25517625Z"}
{"level":"info","msg":"Executing rescan command on device (/sys/block/sdp)","time":"2024-10-04T19:11:22.255188882Z"}
{"level":"debug","msg":"Rescan output","output":"","time":"2024-10-04T19:11:22.798653761Z"}
{"level":"info","msg":"Successful rescan on device (/sys/block/sdp)","time":"2024-10-04T19:11:22.798716914Z"}
{"level":"debug","msg":"Multipath resize output","output":"ok\n","time":"2024-10-04T19:11:22.813356218Z"}
{"level":"info","msg":"Filesystem on mpathfe resized successfully","time":"2024-10-04T19:11:22.813397917Z"}
{"level":"info","msg":"Ext fs: Device /dev/mapper/mpathfe resized successfully","time":"2024-10-04T19:11:42.427917154Z"}
```

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -1298,24 +1315,58 @@ func (s *Service) ControllerExpandVolume(ctx context.Context, req *csi.Controlle
...
+		}
+
 		if vol.Size < requiredBytes {
-			_, err = s.Arrays()[arrayID].Client.ModifyVolume(context.Background(), &gopowerstore.VolumeModify{Size: requiredBytes}, id)
+			_, err = pauseSessionIfMetro(isMetroSession, powerStoreClient, ctx, replicationSessionId)
+			if err != nil {
+				return nil, err
+			}
+
+			_, err = powerStoreClient.ModifyVolume(context.Background(), &gopowerstore.VolumeModify{Size: requiredBytes}, id)
```
Comment (lukeatdell) [inline pkg/controller/controller.go:1352]: should probably log this error if any exists, in case we also hit `resumeSessionErr` and don't get a chance to return this error.
Or combine and return both errors.

Comment (aqu-dell) [reply pkg/controller/controller.go:1352]: Updated

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -1298,24 +1315,58 @@ func (s *Service) ControllerExpandVolume(ctx context.Context, req *csi.Controlle
...
+			}
+			if len(vgs.VolumeGroup) == 0 {
+				replicationSessionId = vol.MetroReplicationSessionID
+			} else {
+				vg := vgs.VolumeGroup[0]
+				rs, err := powerStoreClient.GetReplicationSessionByLocalResourceID(ctx, vg.ID)
+				if err != nil {
+					return nil, err
+				}
+				replicationSessionId = rs.ID
```
[the remark this answers is not in the document]
Comment (aqu-dell) [reply pkg/controller/controller.go:1342]: Updated

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -1298,24 +1315,58 @@ func (s *Service) ControllerExpandVolume(ctx context.Context, req *csi.Controlle
...
+			if err != nil {
+				return nil, err
+			}
+			if len(vgs.VolumeGroup) == 0 {
+				replicationSessionId = vol.MetroReplicationSessionID
+			} else {
+				vg := vgs.VolumeGroup[0]
+				rs, err := powerStoreClient.GetReplicationSessionByLocalResourceID(ctx, vg.ID)
+				if err != nil {
+					return nil, err
```
[the remark this answers is not in the document]
Comment (aqu-dell) [reply pkg/controller/controller.go:1340]: Updated

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -1298,24 +1315,58 @@ func (s *Service) ControllerExpandVolume(ctx context.Context, req *csi.Controlle
 		return nil, status.Errorf(codes.OutOfRange, "volume exceeds allowed limit")
 	}
 
+	powerStoreClient := s.Arrays()[arrayID].Client
```
[the remark this answers is not in the document]
Comment (aqu-dell) [reply pkg/controller/controller.go:1318]: Updated

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -1286,9 +1286,26 @@ func (s *Service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReque
...
+		return powerStoreClient.ExecuteActionOnReplicationSession(ctx, id, "pause", params)
+	}
+
+	return "", nil
+}
+
+func resumeSessionIfMetro(isMetroSession bool, powerStoreClient gopowerstore.Client, ctx context.Context, id string) (gopowerstore.EmptyResponse, error) {
+	if isMetroSession {
+		var params *gopowerstore.FailoverParams
+		powerStoreClient.ExecuteActionOnReplicationSession(ctx, id, "resume", params)
```
[the remark this answers is not in the document]
Comment (aqu-dell) [reply pkg/controller/controller.go:1301]: Updated

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -1286,9 +1286,26 @@ func (s *Service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReque
 	}, nil
 }
 
+func pauseSessionIfMetro(isMetroSession bool, powerStoreClient gopowerstore.Client, ctx context.Context, id string) (gopowerstore.EmptyResponse, error) {
+	if isMetroSession {
+		var params *gopowerstore.FailoverParams
+		return powerStoreClient.ExecuteActionOnReplicationSession(ctx, id, "pause", params)
```
[the remark this answers is not in the document]
Comment (aqu-dell) [reply pkg/controller/controller.go:1292]: Updated

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -1351,9 +1351,34 @@ func (s *Service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReque
 	}, nil
 }
 
+// Returns the metro replication session ID for the volume.
+// If Metro is configured on the volume group that the volume is part of, then returns the ID of the volume group's session.
+func GetMetroSessionID(ctx context.Context, client gopowerstore.Client, volumeID string, volume gopowerstore.Volume, remoteArrayID string) (string, error) {
+	metroSessionID := ""
+	if remoteArrayID != "" {
```
Comment (lukeatdell) [inline pkg/controller/controller.go:1358]: Invert the logic here to return an error if the remote array ID is empty.
as it is currently, it will just return an empty string and no error. This also allows you to bring all the functionality out of the `if` statement

Comment (aqu-dell) [reply pkg/controller/controller.go]: Updated!

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -1363,24 +1388,53 @@ func (s *Service) ControllerExpandVolume(ctx context.Context, req *csi.Controlle
...
+		vol, err := client.GetVolume(ctx, id)
 		if err != nil {
-			return nil, status.Errorf(codes.OutOfRange, "detected SCSI protocol but wasn't able to fetch the volume info")
+			return nil, status.Error(codes.NotFound, "detected SCSI protocol but wasn't able to fetch the volume info")
 		}
+
 		if vol.Size < requiredBytes {
-			_, err = s.Arrays()[arrayID].Client.ModifyVolume(context.Background(), &gopowerstore.VolumeModify{Size: requiredBytes}, id)
+			// get replication session ID if Metro is configured. Pause the session and resume after updating size.
+			replicationSessionID, err := GetMetroSessionID(ctx, client, id, vol, remoteArrayID)
```
Comment (lukeatdell) [inline pkg/controller/controller.go:1407]: `vol.MetroReplicationSessionID` should already be populated here.

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -1363,24 +1388,53 @@ func (s *Service) ControllerExpandVolume(ctx context.Context, req *csi.Controlle
...
 			if err != nil {
-				return nil, err
+				return nil, status.Error(codes.Internal, "error while identifying metro configuration on volume")
+			}
+
+			if replicationSessionID != "" {
```
Comment (lukeatdell) [inline pkg/controller/controller.go:1412]: with the change to `vg.MetroReplicationSessionID` we can just check for an empty string this at the top of the `if` block and return an error if empty. We should also be able to get rid of any additional checks further down

File: pkg/controller/controller_test.go [hunk 10]
```diff
@@ -2267,6 +2268,41 @@ var _ = ginkgo.Describe("CSIControllerService", func() {
...
+				clientMock.On("GetVolume", mock.Anything, validBaseVolID).Return(gopowerstore.Volume{
+					MetroReplicationSessionID: validSessionID,
+					Size:                      validVolSize,
+				}, nil)
+				clientMock.On("ModifyVolume",
+					mock.Anything,
+					mock.AnythingOfType("*gopowerstore.VolumeModify"),
+					validBaseVolID).
+					Return(gopowerstore.EmptyResponse(""), nil)
+				clientMock.On("GetVolumeGroupsByVolumeID", mock.Anything, validBaseVolID).Return([]gopowerstore.VolumeGroup{}, nil)
```
Comment (santhoshatdell) [inline pkg/controller/controller_test.go:2281]: We do not need VG mocks in expand.

**Review by santhoshatdell — COMMENTED** [review]
Can we add more tests coverage the negative scenarios to increase code coverage?

File: pkg/controller/controller_test.go (hunk 10 above)
Comment (aqu-dell) [reply pkg/controller/controller_test.go]: Ah yeah that was mocked for our older version of the code. Updated now

**Review by lukeatdell — APPROVED** [review]

**Review by santhoshatdell — APPROVED** [review]

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -586,8 +579,11 @@ func (s *Service) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest
 			// TODO: Maybe adding volumegroup id/name to volume id can help?
 			_, err := arr.GetClient().RemoveMembersFromVolumeGroup(ctx, &gopowerstore.VolumeGroupMembers{VolumeIDs: []string{id}}, vgs.VolumeGroup[0].ID)
 			if err != nil {
-				// TODO: check for idempotency cases
-				return nil, err
+				if apiError, ok := err.(gopowerstore.APIError); ok && apiError.VolumeAlreadyRemovedFromVolumeGroup() { // idempotency check
+					log.Debugf("Volume %s has already been removed from volume group %s", id, vgs.VolumeGroup[0].ID) // continue to delete volume
```
Comment (donatwork) [inline pkg/controller/controller.go:583]: Is this code for volume group needed?

File: pkg/controller/controller.go [hunk 12]
```diff
@@ -1351,9 +1349,68 @@ func (s *Service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReque
...
+		return false, fmt.Errorf("could not get metro replication session %s", metroSessionID)
+	}
+
+	// confirm the session is in a state we can pause from.
+	if metroSession.State != gopowerstore.RsStateOk &&
+		metroSession.State != gopowerstore.RsStateSynchronizing &&
+		metroSession.State != gopowerstore.RsStatePaused &&
+		metroSession.State != gopowerstore.RsStateSystemPaused &&
+		metroSession.State != gopowerstore.RsStateFractured {
+		return false, fmt.Errorf("could not pause the metro replication session, %s, because the session is not in expected state to pause", metroSession.ID)
```
Comment (donatwork) [inline pkg/controller/controller.go:1364]: What happens here? Does K8s retry the resize after a while and until the volume is expanded? What about the impact to say kubectl edit ...?

File: pkg/controller/controller.go [hunk 13]
```diff
@@ -1351,9 +1349,68 @@ func (s *Service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReque
...
+		metroSession.State == gopowerstore.RsStateResuming ||
+		metroSession.State == gopowerstore.RsStateSwitchingToMetroSync ||
+		metroSession.State == gopowerstore.RsStateFractured {
+		log.Debugf("metro replication session, %s, already resumed", metroSession.ID)
+		return false, nil
+	}
+
+	// metro session can only be resumed if it is in 'paused' state
+	if metroSession.State != gopowerstore.RsStatePaused {
+		return false, errors.New("the metro session must be in 'paused' state before resuming")
```
Comment (donatwork) [inline pkg/controller/controller.go:1399]: Is this the way this is done in other places in the driver? If the desired state is the current state then error? Does the caller check the boolean and do something different?

**Review by donatwork — APPROVED** [review]
Make sure to document that e2e tests were developed and run.

File: pkg/controller/controller.go (hunk 11 above)
Comment (santhoshatdell) [reply pkg/controller/controller.go:583]: Yes, for sync/async mode where VG is involved.

File: pkg/controller/controller.go (hunk 12 above)
Comment (santhoshatdell) [reply pkg/controller/controller.go:1364]: As we noticed, K8s retries the expand operation until it receives success response.
We added some safety checks based on API doc and to cover idempotency case.

File: pkg/controller/controller.go (hunk 13 above)
Comment (santhoshatdell) [reply pkg/controller/controller.go:1399]: The session could be in states other than 'pause' where the API may fail. We are just checking that and failing before-hand. 
That boolean is just for its caller to return appropriate success response back to K8s.

File: pkg/controller/controller.go (hunk 12 above)
Comment (lukeatdell) [reply pkg/controller/controller.go:1364]: Correct, if any call fails for whatever reason during the csi ControllerExpandVolume call, kubernetes will retry the operation.
This check is to make sure the metro session is in a state that can be paused. There are are a few scenarios in which it would not be allowed.

Not sure I'm following on impact related to kubectl edit
